### ISSUE TITLE
Comments are not detected when dealing with decorated classes/functions

### ISSUE THREAD

**Issue by 43081j** [eslint/eslint#11713]
Given the following:

```ts
/**
 * foo
 */
@someDecorator
class Foo {}
```

Both `getJSDocComments` and `getComments` return nothing, as they have a very simple implementation, check if the previous token is a comment block:

https://github.com/eslint/eslint/blob/af81cb3ecc5e6bf43a6a2d8f326103350513a1b8/lib/util/source-code.js#L246

What's the suggested solution here?

Should these methods understand decorators may be present? or is there another way of getting at them already?

Keep in mind estree has no clue what a decorator is i guess, its handled by the appropriate parser (ts, babel, etc). So we can't really have knowledge of them in eslint core.

**Comment by platinumazure** [comment]
Hi @43081j. It's not perfect, but have you tried using the `filter` arguments to the SourceCode methods?

For example, `context.getSourceCode().getPreviousToken(node, { includeComments: true, filter: t => t.type === "Line" || t.type === "Block" })` should give you the comment before the decorator, as long as there is no comment between the decorator and the method. (Obviously some tweaking is needed, but that should get you moving in the right direction.)

**Comment by 43081j** [comment]
@platinumazure i did consider that but wasn't sure how to guard against a situation like so:

```ts
/** foo */ // <- first block comment
let x;

class MyClass {} // <- current node
```

**Comment by kaicataldo** [comment]
@43081j Assuming you're using babel-eslint (which can parse decorators) and that you're starting from the `class` keyword token, could you use `sourceCode#getTokenBefore()` to:
1. Check the token right before the `class` keyword to see if it's a token with `label: "name"`
1. Check the token right before that to see if it's a token with `label: "@"`
1. If so, check the token before that to see if it's a comment

```json
 "tokens": [
    {
      "type": {
        "label": "@",
        "beforeExpr": false,
        "startsExpr": false,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null,
        "updateContext": null
      },
      "start": 0,
      "end": 1,
      "loc": {
        "start": {
          "line": 1,
          "column": 0
        },
        "end": {
          "line": 1,
          "column": 1
        }
      }
    },
    {
      "type": {
        "label": "name",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "value": "someDecorator",
      "start": 1,
      "end": 14,
      "loc": {
        "start": {
          "line": 1,
          "column": 1
        },
        "end": {
          "line": 1,
          "column": 14
        }
      }
    },
    {
      "type": {
        "label": "class",
        "keyword": "class",
        "beforeExpr": false,
        "startsExpr": true,
        "rightAssociative": false,
        "isLoop": false,
        "isAssign": false,
        "prefix": false,
        "postfix": false,
        "binop": null
      },
      "value": "class",
      "start": 15,
      "end": 20,
      "loc": {
        "start": {
          "line": 2,
          "column": 0
        },
        "end": {
          "line": 2,
          "column": 5
        }
      }
    }
]
```

Alternatively, it looks like babel's parser attaches a `decorators` property to `ClassDeclaration` nodes that you could use. If `decorators.length > 0`, you can get the first decorator node and use `sourceCode#getTokenBefore()` with that node to check for the presence of a comment.

```json
 {
        "type": "ClassDeclaration",
        "start": 0,
        "end": 27,
        "loc": {
          "start": {
            "line": 1,
            "column": 0
          },
          "end": {
            "line": 2,
            "column": 12
          }
        },
        "decorators": [
          {
            "type": "Decorator",
            "start": 0,
            "end": 14,
            "loc": {
              "start": {
                "line": 1,
                "column": 0
              },
              "end": {
                "line": 1,
                "column": 14
              }
            },
            "expression": {
              "type": "Identifier",
              "start": 1,
              "end": 14,
              "loc": {
                "start": {
                  "line": 1,
                  "column": 1
                },
                "end": {
                  "line": 1,
                  "column": 14
                },
                "identifierName": "someDecorator"
              },
              "name": "someDecorator"
            }
          }
        ],
        "id": {
          "type": "Identifier",
          "start": 21,
          "end": 24,
          "loc": {
            "start": {
              "line": 2,
              "column": 6
            },
            "end": {
              "line": 2,
              "column": 9
            },
            "identifierName": "Foo"
          },
          "name": "Foo"
        },
        "superClass": null,
        "body": {
          "type": "ClassBody",
          "start": 25,
          "end": 27,
          "loc": {
            "start": {
              "line": 2,
              "column": 10
            },
            "end": {
              "line": 2,
              "column": 12
            }
          },
          "body": []
        }
      }
```
